source-snowflake/discovery.go

```diff
@@ -0,0 +1,383 @@
+					"properties": map[string]*jsonschema.Schema{
+						metadataProperty: {
+							Type: "object",
+							Extras: map[string]interface{}{
+								"properties": map[string]*jsonschema.Schema{
+									"op": {
+										Enum:        []interface{}{"c", "d", "u"},
+										Description: "Change operation type: 'c' Create/Insert, 'u' Update, 'd' Delete.",
+									},
+									"source": sourceSchema,
+									"before": {
+										Ref:         "#" + anchor,
+										Description: "Record state immediately before this change was applied.",
+										Extras: map[string]interface{}{
+											"reduce": map[string]interface{}{
+												"strategy": "firstWriteWins",
+											},
+										},
+									},
+								},
+								"reduce": map[string]interface{}{
+									"strategy": "merge",
+								},
+							},
+							Required: []string{"op", "source"},
+						},
+					},
+					"reduce": map[string]interface{}{
+						"strategy": "merge",
+					},
+				},
+				Required: []string{metadataProperty},
+			},
+			{Ref: "#" + anchor},
+		},
+	}
+
+	var documentSchema, err = schema.MarshalJSON()
+	if err != nil {
+		return nil, fmt.Errorf("error marshalling schema JSON: %w", err)
+	}
+	return documentSchema, nil
+}
+
+func translateDBToJSONType(column *snowflakeDiscoveryColumn) (*jsonschema.Schema, error) {
+	var schema columnSchema
+	switch column.DataType {
+	case "NUMBER":
+		if column.NumericScale != nil && *column.NumericScale == 0 {
+			schema = columnSchema{jsonType: "integer"}
+		} else {
+			schema = columnSchema{jsonType: "number"}
+		}
+	default:
+		if s, ok := snowflakeTypeToJSON[column.DataType]; ok {
+			schema = s
+		} else {
+			return nil, fmt.Errorf("unhandled Snowflake type %q (found on column %q of table %q)", column.DataType, column.Name, column.Table)
+		}
+	}
+
+	// Pass-through the column description and nullability.
+	schema.nullable = column.IsNullable != "NO"
+	if column.Comment != nil {
+		schema.description = *column.Comment
+	}
+	return schema.toType(), nil
+}
+
+type columnSchema struct {
+	contentEncoding string
+	description     string
+	format          string
+	nullable        bool
+	extras          map[string]interface{}
+	jsonType        string
+}
+
+func (s columnSchema) toType() *jsonschema.Schema {
+	var out = &jsonschema.Schema{
+		Format:      s.format,
+		Description: s.description,
+		Extras:      make(map[string]interface{}),
+	}
+	for k, v := range s.extras {
+		out.Extras[k] = v
+	}
+
+	if s.contentEncoding != "" {
+		out.Extras["contentEncoding"] = s.contentEncoding // New in 2019-09.
+	}
+
+	if s.jsonType == "" {
+		// No type constraint.
+	} else if s.nullable {
+		out.Extras["type"] = []string{s.jsonType, "null"} // Use variadic form.
+	} else {
+		out.Type = s.jsonType
+	}
+	return out
+}
+
+var snowflakeTypeToJSON = map[string]columnSchema{
+	// "NUMBER":  {jsonType: "number"}, // The 'NUMBER' column type is handled in code
+	"TEXT":          {jsonType: "string"},
```

Comment:
I think we'll want to handle `VARIANT` types also. Maybe the "any possible type" schema is best for them still, but I think they are fairly common and not something we'd want to show an error for - which if I'm understanding things correctly, we currently would.

Reply:
Variant types (including object/array) are rather tricky to handle. The values we get back from the gosnowflake client library are stringified and appear to be JSON, so it's tempting to translate them all to `json.RawMessage` values so they get embedded directly into the captured documents. However I can't find any actual documentation which would justify doing this. I'll have to investigate this one some more tomorrow morning.

Reply:
The connector now has support for the `VARIANT` / `OBJECT` / `ARRAY` Snowflake column types. Discovery will produce schemas of `{}` / `{type: object}` / `{type: array}` respectively and captures produce the obvious JSON representations.